### PR TITLE
Const indexing and friends

### PR DESCRIPTION
**Pull request by oli-obk**
This PR allows the constant evaluation of index operations on constant arrays and repeat expressions. This allows index expressions to appear in the expression path of the length expression of a repeat expression or an array type.

An example is 

``` rust
const ARR: [usize; 5] = [1, 2, 3, 4, 5];
const ARR2: [usize; ARR[1]] = [42, 99];
```

In most other locations llvm's const evaluator figures it out already. This is not specific to index expressions and could be remedied in the future.

### CODE REVIEW

**Comment by nrc** [comment]
cc @eddyb

**Comment by nrc** [comment]
Given that we're in the post-1.0 world. We need to either make this not a breaking change or justify how this is effectively just fixing a bug (we'd also need to measure the impact of the change in this case). So I don't think we can land this as is.

I'm afraid I don't understand either of the two proposals for fixing the breaking change - could you explain them in a bit more detail please?

**Comment by oli-obk** [comment]
1. introducing a static fallback
if the const evaluation of an indexing operation fails, we can call the old code that did the indexing in llvm
This would be the safe solution, as there might be other situations where static evaluation (llvm) works but constant evaluation doesn't.

2. by making enum variants first class consts
I already created new `const_val` variants for Array, Repeat and Ref. I could add an EnumVariant const val and process it properly. Right now there's just some magic that works if a c-like enum is const evaluated.
This is the nice solution, as it's another step towards full const evaluation of all const-evaluable code.

File: src/librustc/middle/const_eval.rs [hunk 1]
```diff
@@ -877,12 +962,57 @@ pub fn eval_const_expr_partial<'tcx>(tcx: &ty::ctxt<'tcx>,
       ast::ExprBlock(ref block) => {
         match block.expr {
             Some(ref expr) => try!(eval_const_expr_partial(tcx, &**expr, ety)),
-            None => const_int(0)
+            None => const_int(0) // huh???
```
Comment (oli-obk) [inline src/librustc/middle/const_eval.rs]: added in 24ece07cecf3df59ce7b9ed3b317e3ffd718f756

I believe it should be `unreachable!()` or a `sess.span_bug` but I might be wrong

cc @Kimundi

Comment (oli-obk) [inline src/librustc/middle/const_eval.rs]: oh sorry, should have been researching better: https://github.com/rust-lang/rust/pull/14183#discussion_r12612731

**Comment by oli-obk** [comment]
hold this off until #25609 is merged, I'll rebase then.

File: src/librustc_trans/trans/consts.rs [hunk 2]
```diff
@@ -566,12 +610,20 @@ fn const_expr_unadjusted<'a, 'tcx>(cx: &CrateContext<'a, 'tcx>,
           }
 
           ast::ExprIndex(ref base, ref index) => {
+              match const_eval::eval_const_expr_partial(cx.tcx(), &e, None) {
```
Comment (oli-obk) [inline src/librustc_trans/trans/consts.rs]: an alternative would be to check if `base` points to a static or a constant and then decide on what to do. Any thoughts?

**Comment by eddyb** [comment]
I thought constants couldn't refer to statics at all. Am I missing something?

**Comment by oli-obk** [comment]
> I thought constants couldn't refer to statics at all. Am I missing something?

yea, that confused me quite a bit, too. As I read the docs and book, it means that if you use a constant in between, you again loose all guarantees of memory locations and uniqueness of values. Although the compiler actually evaluates all constants with intermediate statics as statics

**Comment by eddyb** [comment]
@oli-obk I realized the code path is for the value with which a `static` is initialized, not a `const`.
A `const` cannot mention any `static`, at all, it's the initializer value of a `static` which can.

Also, the error `constants cannot refer to other statics, insert an intermediate constant instead [E0013]` is very confusing, I'm not sure what it's trying to tell me.

**Comment by oli-obk** [comment]
oh wow, I got everything confused here... I was so sure statics were simply evaluated as constants if called from a constant...

So I guess my "try const eval, if fails try static eval" is fine?

**Comment by oli-obk** [comment]
> Also, the error `constants cannot refer to other statics, insert an intermediate constant instead [E0013]` is very confusing, I'm not sure what it's trying to tell me.

It's trying to tell you you should replace the static with a constant, and make the static simply be a `static A: T = C;` where `C` is your new constant. Then your other constant that failed with the error can also reference `C` instead of `A`.

**Comment by oli-obk** [comment]
rebased, it was only a minor change (forwarding a function parameter). `make check` is running

**Comment by eddyb** [comment]
@oli-obk that seems very contrived, in practice you replace the referenced `static` with a `const` or you replace the referring `const` with a `static`.

**Comment by oli-obk** [comment]
I'd rather remove the error completely and have constant evaluation treat statics as constants. But then my "try const eval, else fallback to static eval" method fails horribly.

`check-stage1` passes

**Comment by eddyb** [comment]
@oli-obk No, no, no: statics **are not** constant.
The main two usecases for `static` are interior mutability and link-time constants.
Neither of these can be given meaningful semantics when used in a constant.
IMO a `static` which could be `const` should be linted against.

**Comment by oli-obk** [comment]
now that is something i can do! useless static-lint is on my TODO list

**Comment by oli-obk** [comment]
rebased and removed the const references

This PR now only addresses const arrays, repeats and indexing into them.

File: src/librustc/middle/const_eval.rs [hunk 3]
```diff
@@ -357,11 +359,15 @@ pub enum ErrKind {
...
     NegateOnTuple,
+    NegateOnArray,
+    NegateOnRepeat,
     NotOnFloat,
     NotOnString,
     NotOnBinary,
     NotOnStruct,
     NotOnTuple,
+    NotOnArray,
+    NotOnRepeat,
```
Comment (eddyb) [inline src/librustc/middle/const_eval.rs]: This feels like it should be a nested enum or something (is it done like this just for flattening?).

File: src/test/compile-fail/array_const_index-0.rs [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+static A: &'static [i32] = &[];
+static B: i32 = (&A)[1]; //~ ERROR: const index-expr is out of bounds
```
Comment (eddyb) [inline src/test/compile-fail/array_const_index-0.rs]: Why is this allowed at all?!

**Comment by eddyb** [comment]
I just checked and it seems that many of these cases just cause the compiler to abort (with SIGILL and no error message), for example ([playpen](https://play.rust-lang.org/?gist=90940f28028322c06b01&version=stable)):
```rust
fn main() {
    static A: i32 = 0;
    static B: i32 = *&A;
    println!("{:?}", B);
}
```
Though it only happens when the statics are inside a function, so I'm a bit confused as to what's going on.

IMO references to statics should not be dereferenced for any other purpose than to offset them (similar to `&(*(0 as *const Struct)).field as *const _ as usize`, which is no longer allowed).

Is it even possible to deny such uses after 1.0?
cc @alexcrichton @nikomatsakis

File: src/test/compile-fail/array_const_index-0.rs (hunk 4 above)
Comment (oli-obk) [inline src/test/compile-fail/array_const_index-0.rs]: that was allowed before my change, I just added the test

Comment (oli-obk) [inline src/test/compile-fail/array_const_index-0.rs]: https://github.com/rust-lang/rust/issues/3170#issuecomment-24050516

**Comment by oli-obk** [comment]
so... I came to the conclusion that indexing into and derefing statics never makes sense in a static environment. except when taking the address directly again, but that's apparently not ok:

>Borrowed statics can specifically *only* have their address taken, not any number of other borrows such as borrowing fields, reading elements of an array, etc.

taking either of these by value also makes no sense... so i tried to forbid that, but this also breaks constants since there's no difference between a `&'static` that points to a constant and one that points to a static... so... one solution is to introduce `&'const` (even if just in the `ty::Region` enum), other solutions are welcome.

**Comment by oli-obk** [comment]
rebased

**Comment by oli-obk** [comment]
@eddyb should this PR be on hold until the static deref issue is fixed since it'd introduce yet another way to do static deref?

File: src/librustc/middle/check_const.rs [hunk 5]
```diff
@@ -813,8 +815,25 @@ impl<'a, 'tcx> euv::Delegate<'tcx> for CheckCrateVisitor<'a, 'tcx> {
...
+                }
 
+                mc::cat_rvalue(ty::ReStatic) if is_interior => {
+                    match self.mode {
+                        Mode::Static | Mode::StaticMut => {
+                            span_err!(
+                                self.tcx.sess,
+                                consume_span,
+                                E0398,
+                                "cannot dereference a &'static T",
```
Comment (eddyb) [inline src/librustc/middle/check_const.rs]: I'm not sure I understand this bit... it would disallow dereferencing a `&'static T` inside a `static`? What good would that do?

Comment (oli-obk) [inline src/librustc/middle/check_const.rs]: in case you have

```rust
static A: i32 = 5;
static B: i32 = *&A;
```

which is forbidden, but

```rust
static A: i32 = 5;
let x = *&A;
```

should be allowed

but as I said. this also breaks dereferencing references to constants

**Comment by pnkfelix** [comment]
@oli-obk Okay

 1. Sorry it has taken me so long to get around to reading your PR

 2. Overall I think the code itself, as of the current PR, is pretty clean and easy to follow, so that's good

 3. I am concerned about the potential for breaking changes to the language as discussed by @nrc earlier in this issue.  Based on the massive amount of text that you have crossed out of the description, I infer that this is *not* a breaking change anymore.  Is this correct?

 4. Assuming that this is no longer a breaking change, I personally would like it if the description were revised to avoid so much long text that is implicitly crossed out -- when bors commits this PR, it is going to be too hard for readers to parse the text of those commits when they are not rendered by markdown.  I understand the desire to preserve the history of the description; I suggest revising the description to a format like this:

   ```
This is the new description.  It is short and reflects the 
state of the PR at the time it is landing.

(Old description follows)

----

Here you ~~cut-and-paste~~ transcribe the current
description that ~~pnkfelix~~ someone complained
about with ~~lots of~~ embedded strike-outs.
Hopefully the reader using git in a terminal will
understand that they should not bother looking too
much at this text.
   ```

Personally I'm actually relatively okay with the idea of merging the code changes proposed in this PR, assuming my understanding from above that this is a non-breaking change is correct.

@nrc do you have thoughts on this?

One last thought: We *might* consider adding a feature-gate to try cover our collective asses, i.e. give the change some time to breathe without risk of breakage leaking into the stable world.  But I'm not sure a feature gate pays for itself here.

File: src/librustc/middle/const_eval.rs (hunk 3 above)
Comment (pnkfelix) [inline src/librustc/middle/const_eval.rs]: (I'm okay with not trying to fold that particular change into this PR -- i.e. switching to a nested enum might make sense but also seems like it is a broader refactoring that we could develop later, no?)

**Comment by oli-obk** [comment]
>  I personally would like it if the description were revised to avoid so much long text that is implicitly crossed out 

done

> We might consider adding a feature-gate to try cover our collective asses

I have removed the potentially critical part from trans (it was more of an ugly hack anyway). Now only code in the const evaluator is touched.

**Comment by oli-obk** [comment]
**DO NOT MERGE**: as noted in https://github.com/rust-lang/rust/pull/26848#issuecomment-126126181 this is a breaking change since it can turn runtime-assertions into compile-time errors.

**Comment by steveklabnik** [comment]
So, what's up with this PR? given that it's breaking, does that mean we're not taking it?

**Comment by nrc** [comment]
cc @rust-lang/lang 

If we're sure that the only new compile errors are where we would previously have had a run-time error and we get a clean crater run, then I'd be happy to merge this. What about the rest of the lang team?

@oli-obk any new thoughts?

**Comment by oli-obk** [comment]
I was on vacation. Will take care of #26848 and this PR in the next week.

**Comment by nikomatsakis** [comment]
I believe the general consensus was that we should alter the approach to constant evaluation to make this not a breaking change, right?

**Comment by oli-obk** [comment]
This is not a breaking change anymore. I also rebased this PR. It is ready to be reviewed again.

File: src/librustc/middle/const_eval.rs [hunk 6]
```diff
@@ -1092,11 +1108,71 @@ pub fn eval_const_expr_partial<'tcx>(tcx: &ty::ctxt<'tcx>,
       hir::ExprBlock(ref block) => {
         match block.expr {
             Some(ref expr) => try!(eval_const_expr_partial(tcx, &**expr, ty_hint, fn_args)),
-            None => Int(0)
+            None => unreachable!(),
```
Comment (oli-obk) [inline src/librustc/middle/const_eval.rs]: this is unrelated, but I was not able to trigger this code-path, so I marked it as unreachable

**Comment by oli-obk** [comment]
uhm... I'm getting a stackoverflow in `[run-pass] run-pass/match-arm-statics.rs`. I'm not sure what's going on. The last time I ran the tests this didn't happen...

**Comment by oli-obk** [comment]
travis doesn't see the stack overflow, maybe I had some debug flags turned on... I'm trying again on my machine. But since travis likes it, this is ready to be merged.

**Comment by oli-obk** [comment]
rebased

**Comment by bluss** [comment]
The PR's message will go into the git log when this is merged. I think it's best to make sure it reflects the merged version, and I guess it's best to remove the outdated text?

**Comment by oli-obk** [comment]
rebased.

@bluss, I moved it to my second comment

**Comment by bluss** [comment]
@oli-obk Thank you, that's neater!

File: src/librustc/middle/const_eval.rs [hunk 7]
```diff
@@ -416,6 +424,12 @@ pub enum ErrKind {
     ExpectedConstTuple,
     ExpectedConstStruct,
     TupleIndexOutOfBounds,
+    IndexedNonVec,
+    IndexNotNatural,
```
Comment (nikomatsakis) [inline src/librustc/middle/const_eval.rs]: Nit: I feel like I vaguely remember what "natural numbers" are from grade-school, but wouldn't "positive" be easier?

File: src/test/run-pass/array_const_index-1.rs [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+
+fn main() {
+    const ARR: [i32; 6] = [42, 43, 44, 45, 46, 47];
+    const IDX: usize = 3;
+    const VAL: i32 = ARR[IDX];
+    const BLUB: [i32; (ARR[0] - 41) as usize] = [5];
```
Comment (nikomatsakis) [inline src/test/run-pass/array_const_index-1.rs]: Silly maybe, but you could add a compile-fail test where the type evaluates to the *wrong* length? (Just to show we're doing the arithmetic properly)

**Comment by nikomatsakis** [comment]
So this PR looks good to me.

One question is whether we should feature-gate these changes? Otherwise it does represent an implicit commitment. OTOH, it seems like something we'll obviously want to support. I feel like we've chosen to feature-gate in other similar instances. But it seems like we need to make better procedure for moving forward with UNGATING. I know that @aturon made a bunch of issues, which is a first-step, we need to start reviewing them regularly as well, I imagine. Thoughts @rust-lang/lang?

**Comment by pnkfelix** [comment]
@nikomatsakis I suppose the broader question is what criteria we should use in deciding whether to impose a feature gate.

I see three potential reasons for a feature gate:

 1. When we do not know whether we want a feature (in any form) at all

 2. When we have decided we want a feature, but there are details about its specified semantics that we are not sure about. (Note that there a range of potential semantics that we can assign, and there is a partial ordering between them, since we can reject code statically when it fails to adhere to some base set of rules ... so we might feature-gate some extension to a semantics, but leave the core ungated).

 3. When we think we know the semantics we want for a feature, but we are not confident that we can commit to the "current" implementation.

Does that sounds about right?

In this case, I *think* the answers are: (1.) we want this (but maybe others would disagree), (2.) any possible semantics we might choose is going to include what is outlined here, I think (i.e. this probably occupies a low point in the partial ordering of potential semantics that I alluded to above). So the only question mark for me is (3.): is the implementation sufficiently low-risk that we can side-step a gate?  And I don't have a confident answer there.

File: src/librustc/middle/const_eval.rs (hunk 7 above)
Comment (pnkfelix) [inline src/librustc/middle/const_eval.rs]: Positive doesn't include zero. (Usually the naturals start at zero.)

You could do `IndexNotNonNegative`, or just `IndexNegative` if you like.  :)

Comment (pnkfelix) [inline src/librustc/middle/const_eval.rs]: (though perhaps this error is also issues when the index is a floating point number, i.e. a non-integer?  That's the other reason one would say "Natural" ...)

Comment (oli-obk) [inline src/librustc/middle/const_eval.rs]: I like `IndexNegative`, I'll go with that.

Natural numbers only include integers.

**Comment by oli-obk** [comment]
We already allow indexing into constants anywhere but a true constant (e.g. array lengths). It's just computed by llvm's constant evaluator in `trans/consts.rs` So this is more of a bugfix ;)

So imo it's only a question about (3.)

**Comment by oli-obk** [comment]
addressed nits

**Comment by nikomatsakis** [comment]
So let me be clear on my concern. In general, I think that constant expression evaluation is a kind of "ad-hoc, best-effort" mess whose limits are not well documented and are only understood by a relatively small set. The very fact that we have two constant evaluators that handle two distinct subsets -- and slightly differently, at times, though of course those are "just bugs" -- is sort of what I'm talking about. I would very much like to refactor things so that we have ONE constant evaluator that is following a relatively documented, comprehensible process. And I'm wary of accidentally enabling evaluation of expressions that turns out to work most of the time, but have rough edges. Let me give you another concrete example: we currently allow you to "dereference" an `&'static` pointer, which -- when it points into a `static` -- perhaps ought to be disallowed. Certainly it leads to https://github.com/rust-lang/rust/issues/29884, which was partly because of a (rather ad-hoc) decision to support static recursion (https://github.com/rust-lang/rust/issues/29719). Luckily we put a feature-gate on this decision, so we have some luxury to either prohibit derefs into statics or else fix https://github.com/rust-lang/rust/issues/29719 before we are overly committed. So I think it's pretty reasonable to move slow in terms of allowing arbitrary stable code to make use of extended sets of constant expressions.

Now, indexing and dereferences are basically the same thing, so an example of something I'd like to be careful about is: what happens if you try to index into the content of a `static` here?

All this is not to say we need a feature-gate, but just to say that there is evidence that there is reason to think that @pnkfelix's condition #2 may apply here:

> When we have decided we want a feature, but there are details about its specified semantics that we are not sure about.

Basically, I think it might be prudent to place a kind of feature gate around ALL expansions of constants until we have something a bit more disciplined (at least a unified evaluator, ideally an RFC laying out what is expected to work).

**Comment by nikomatsakis** [comment]
I just want to be clear though: I'm also open to the idea that I'm being excessively cautious. Part of my problem I think is that **I personally** don't feel like I deeply understand the limits of the current system, and that makes me feel uneasy. Perhaps those more intimately familiar feel more comfortable.

**Comment by oli-obk** [comment]
> Now, indexing and dereferences are basically the same thing, so an example of something I'd like to be careful about is: what happens if you try to index into the content of a static here?

you get an error, because a static is not a true constant and `const_eval` aborts.

>  would very much like to refactor things so that we have ONE constant evaluator that is following a relatively documented, comprehensible process. 

This is the target I've been doing all these PRs for. Once `const_eval` can do anything `trans/consts.rs` can, I want to change `trans/consts.rs` to be nothing but a translator for instances of the `ConstVal` enum.

> All this is not to say we need a feature-gate, but just to say that there is evidence that there is reason to think that @pnkfelix's condition #2 may apply here:

I'm perfectly fine with a feature gate.

**Comment by oli-obk** [comment]
added a feature gate. make check passes.
